simulate/simulate.go

```diff
@@ -121,166 +132,360 @@ func changeBase(originalBase dna.Base) dna.Base {
+				codon.Seq = make([]BaseExt, 3)
+				if p%3 == 0 {
+					codon.Seq[0] = exonSeq[p]
+					codon.Seq[1] = exonSeq[p+1]
+					codon.Seq[2] = exonSeq[p+2]
+					allCodons = append(allCodons, codon)
+					//log.Print(allCodons)
+				}
+			}
+		} else if gene.ExonFrames[exon] == 0 && gene.ExonFrames[exon+1] != 0 { //Frame of this exon = 0 but Frame of next exon != 0
+			log.Print("next Exon's frame !=0")
+			log.Print(gene.ExonFrames)
+			log.Print(gene)
+			endFrame := gene.ExonFrames[exon+1]
+			nextExonStart := gene.ExonStarts[exon+1]
+			exonSeq = seq[exonStart:exonEnd]
+			for ef := nextExonStart; ef < nextExonStart+endFrame; ef++ {
+				exonSeq = append(exonSeq, seq[ef])
+			}
+			for p := 0; p < len(exonSeq); p++ {
+				var codon CodonExt
+				codon.Seq = make([]BaseExt, 3)
+				if p%3 == 0 {
+					codon.Seq[0] = exonSeq[p]
+					codon.Seq[1] = exonSeq[p+1]
+					codon.Seq[2] = exonSeq[p+2]
+					allCodons = append(allCodons, codon)
+				}
+			}
+		} else if gene.ExonFrames[exon] != 0 && gene.ExonFrames[exon+1] != 0 { //Frame of the last exon and this exon != 0
+			log.Print("this != 0 next != 0")
+			startFrame := gene.ExonFrames[exon]
+			endFrame := gene.ExonFrames[exon+1]
+			nextExonStart := gene.ExonStarts[exon+1]
+			exonSeq = seq[(gene.ExonStarts[exon] + startFrame):gene.ExonEnds[exon]]
+			for ef := nextExonStart; ef < nextExonStart+endFrame; ef++ {
+				exonSeq = append(exonSeq, seq[ef])
+			}
+			for p := 0; p < len(exonSeq); p++ {
+				var codon CodonExt
+				codon.Seq = make([]BaseExt, 3)
+				if p%3 == 0 {
+					codon.Seq[0] = exonSeq[p]
+					codon.Seq[1] = exonSeq[p+1]
+					codon.Seq[2] = exonSeq[p+2]
+					allCodons = append(allCodons, codon)
+				}
+			}
+		}
+	} else { //this exon is the last exon
+		if gene.ExonFrames[exon] == 0 { //Frame = 0
+			exonSeq = seq[exonStart:exonEnd]
+			for p := 0; p < len(exonSeq); p++ {
+				var codon CodonExt
+				codon.Seq = make([]BaseExt, 3)
+				if p%3 == 0 {
+					codon.Seq[0] = exonSeq[p]
+					codon.Seq[1] = exonSeq[p+1]
+					codon.Seq[2] = exonSeq[p+2]
+					allCodons = append(allCodons, codon)
+				}
+			}
+		} else if gene.ExonFrames[exon] != 0 { //Frame != 0
+			startFrame := gene.ExonFrames[exon]
+			exonSeq = seq[(gene.ExonStarts[exon] + startFrame):gene.ExonEnds[exon]]
+			for p := 0; p < len(exonSeq); p++ {
+				var codon CodonExt
+				codon.Seq = make([]BaseExt, 3)
+				if p%3 == 0 {
+					codon.Seq[0] = exonSeq[p]
+					codon.Seq[1] = exonSeq[p+1]
+					codon.Seq[2] = exonSeq[p+2]
+					allCodons = append(allCodons, codon)
 				}
 			}
 		}
 	}
+	return allCodons
+}
+
+//BasesToBaseExt converts a slice of dna.Base to a slice of BaseExt
+func BasesToBaseExt(seq []dna.Base) []BaseExt {
+	Sequence := make([]BaseExt, len(seq))
+	for i := 0; i < len(seq); i++ {
+		Sequence[i].Base = seq[i]
+		Sequence[i].SeqPos = i
+	}
+	return Sequence
+}
 
+//BaseExtToBases converts a slice of BaseExt to a slice of dna.Base
+func BaseExtToBases(seq []BaseExt) []dna.Base {
```

Comment:
I think this is taking bases that have been added in any order, looks at the position in their struct, and then puts them back in order.  I am thinking that you might be using append(), when you could be sticking things directly where they need to be by making a slice of a given length ( whatever = make([]dna.Base, 100) ) and then placing bases (or anything else) directly in it using: whatever[i] = myBase